### ISSUE TITLE
After canceling a proposal all other attempts to execute a proposal would revert

### ISSUE THREAD

**Issue by c4-submissions** [code-423n4/2023-10-party-findings#501]
# Lines of code

https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/proposals/ProposalExecutionEngine.sol#L146-L204

# Vulnerability details

## Impact

If any proposal gets cancelled all other other proposals wouldn't be executable.

## Proof of Concept

The [`cancelProposal()`](https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/proposals/ProposalExecutionEngine.sol#L207) is used to be forcibly cancel an incomplete proposal, there is the intention of this being last resort, so whenever possible, proposals should be allowed to complete their entire lifecycle, below is it's implementation.

```solidity
function cancelProposal(uint256 proposalId) external onlyDelegateCall {
  // Must be a valid proposal ID.
  if (proposalId == 0) {
    revert ZeroProposalIdError();
  }
  Storage storage stor = _getStorage();
  {
    // Must be the current InProgress proposal.
    uint256 currentInProgressProposalId = stor.currentInProgressProposalId;
    if (currentInProgressProposalId != proposalId) {
      revert ProposalNotInProgressError(proposalId);
    }
  }
  //@audit
  stor.currentInProgressProposalId = 0;
  stor.nextProgressDataHash = 0;
}
```

Issue is that, whenever a proposal gets cancelled, the `current InProgress proposal ID and next progress data` gets cleared, or in other words reset to 0, this entirely breaks the idea in `executeProposal()` since each proposals to be executed are somewhat chained.

Take a look at [ `executeProposal()`](https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/proposals/ProposalExecutionEngine.sol#L146-L204), most especially the initial proposal ID check and assignment:

```solidity
if (params.proposalId == 0) {
    revert ZeroProposalIdError();
}
...
//@audit-info this would be true after  `cancelProposal()` has been executed
if (currentInProgressProposalId == 0) {
    // No proposal is currently in progress.
    stor.currentInProgressProposalId = params.proposalId;
} else if (currentInProgressProposalId != params.proposalId) {
    // Only one proposal can be in progress at a time.
    revert ProposalExecutionBlockedError(params.proposalId, currentInProgressProposalId);
}
```

`if (currentInProgressProposalId == 0)` would be true after `cancelProposal()` has been executed, note that after that this line: `stor.currentInProgressProposalId = params.proposalId` updates the `proposalId` and it's no longer 0.

Below is the subsequent actions taken in `executeProposal()` with regards to `nextProgressDataHash`

```solidity
            bytes32 nextProgressDataHash = stor.nextProgressDataHash;
            //@audit-info This would be true after `cancelProposal()`
            if (nextProgressDataHash == 0) {
                // Expecting no progress data.
                // This is the state if there is no current `InProgress` proposal.
                assert(currentInProgressProposalId == 0);
                if (params.progressData.length != 0) {
                    revert ProposalProgressDataInvalidError(
                        keccak256(params.progressData),
                        nextProgressDataHash
                    );
                }
```

Since `nextProgressDataHash` is zero after a previous proposal has been canceled, the above code executes, but we can see that context here is completely wrongly implemented, cause

- A check is made that `params.progressData.length` should not be zero which would fail, cause `nextProgressDataHash` is attached to every proposal there is and this is used to enforced the chaining of the proposals, i.e the next proposal that's to be executed _MUST_ have it's progressDataHash matching this value. this can be seen to be explained [here](https://github.com/code-423n4/2023-10-party/blob/b23c65d62a20921c709582b0b76b387f2bb9ebb5/contracts/proposals/ProposalExecutionEngine.sol#L63-L82), essentially causing the attempt to execute the next proposal always reverting with the ` ProposalProgressDataInvalidError()`.

## Tool used

Manual Review

## Recommended Mitigation Steps

The functionality of `cancelProposal` if attached with `executeProposal()` needs to be refactored

Though this fix has not been battle tested by me, at one glance it seems the best thing to do in this case, is that whenever a proposal gets cancelled, instead of resetting the stored `nextProgressDataHash` to 0, query the `nextProgressDataHash` that's attached to that specifc `proposalId` and store it as `stor.nextProgressDataHash` instead.

> Edit: At the final minutes of the contest, I gave this a re-read, having forgotten some of the context, I think there is a possibility the function even panic errors out at this line: `assert(currentInProgressProposalId == 0);` since the proposal Id has already been previously updated in the section below:

```solidity
        if (currentInProgressProposalId == 0) {
            // No proposal is currently in progress.
            // Mark this proposal as the one in progress.
            stor.currentInProgressProposalId = params.proposalId;
```

Which still makes this issue valid as all proposals after `cancelProposal()` still revert, cause as explained in the report, after canceling, `currentInProgressProposalId` is set to 0 and since the next call to `execute()` would update the proposal Id before the assertion, `assert(currentInProgressProposalId == 0)` panic errors.

Note that fixes for both issues need to be done, i.e `cancelProposal()` should not reset `nextProgressDataHash` to 0, and this assertion should be removed, since the function currently updates the current proposal Id from 0 to the current one.

TLDR: `cancelProposal()` needs to be _heavily_ refactored.

## Assessed type

Invalid Validation
